api/main.py

```diff
@@ -368,3 +371,120 @@ class parallelItem(BaseModel):
     except KeyError as e:
         print("KeyError: ", e)
         raise HTTPException(status_code=400)
+
+
+@app.get("/visual/{searchterm}")
+async def get_graph_for_file(
+    searchterm: str,
+    language: str,
+    selected: List[str] = Query([]),
+):
+    try:
+        db = get_db()
+        total_parallel_count = []
+        searchtype = 'category'
+        if re.search("^[A-Z][a-z]+$", searchterm):
+            searchtype = "collection"
+
+        # get a sorted list of categories to get the results in the right order
+        query_full_category_list = db.AQLQuery(
+            query=query_sorted_category_list,
+            bindVars={
+                "language": language,
+                "selected": selected,
+            },
+        )
+        category_list = {}
+        for category in query_full_category_list.result:
+            category_list.update(category)
+
+        # check if the search is for a catagory (i.e. T06) or for a collection (i.e. Tengyur)
+        if searchtype == "category":
+            filecount = get_query_files_per_category(language+"_"+searchterm)
+
+            for filename in filecount:
+                parallel_count = filename["totallengthcount"]
+                for categoryname in category_list.keys():
+                    if categoryname in parallel_count.keys():
+                        total_parallel_count.append([filename["filename"],"R_"+categoryname+" "+category_list[categoryname],parallel_count[categoryname]])
+                    else:
+                        total_parallel_count.append([filename["filename"],"R_"+categoryname+" "+category_list[categoryname],0])
+
+        # if the search is for a collection, a list of categories for that collection
+        # is iterated over and the results for each file added.
+        elif searchtype == "collection":
+            query_collection_list = db.AQLQuery(
+                query=query_categories_per_collection,
+                bindVars={
+                    "searchterm": language+"_"+searchterm,
+                    "language": language
+                },
+            )
+
+            for cat, catname in query_collection_list.result[0].items():
+                filecount = get_query_files_per_category(language+"_"+cat)
+
+                total_parlist = {}
+                for filename in filecount:
```

Comment:
This seems overly verbose, but I don't have quick ideas how to simplify it now. I would merge for now if it works fast enough, but keep in mind we need to refactor it later once we have more time.

Reply:
Yes, I have some ideas what to do here.